Adds Fluent Image web component

halerankin opened the pull request:
Adds Image web component.

chrisdholt commented on packages/web-components/src/image/image.template.ts:
```diff
@@ -0,0 +1,23 @@
+ * @public
+ */
+export const template: ElementViewTemplate<Image> = html<Image>`
+  <template>
```
Template elements are not necessary in cases like this - we only would need one here in the case we need to bind data to the host element.

chrisdholt replied:
with that said, as noted below, some of the attributes which aren't specific to `img` should be enumerated here if we were to go this route.

chrisdholt commented on packages/web-components/src/image/image.template.ts:
```diff
@@ -0,0 +1,23 @@
+ */
+export const template: ElementViewTemplate<Image> = html<Image>`
+  <template>
+    <img
+      alt="${x => x.alt}"
+      block="${x => x.block}"
+      border="${x => x.border}"
+      borderRadius="${x => x.borderRadius}"
+      fit="${x => x.fit}"
+      margin="${x => x.margin}"
+      role="${x => (x.presentation ? 'presentation' : null)}"
+      shadow="${x => x.shadow}"
+      shape="${x => x.shape}"
+      src="${x => x.src}"
+    />
```
A few thoughts here - 

- This shadows the element and makes it a requirement for us to enumerate _any_ attribute that needs to go *onto* this image. So, if we need to decorate this with `aria-`, we need to enumerate and create an attribute and corresponding property for every valid aria attribute. This is the same for any valid attribute for a native image, so we need to enumerate the lazy loading attributes as well.
- 1 above is super costly and a lot of JS to add a simple image
- I think a more performant and ergonomic solution for developers here is to let the platform do the work. The intent of the image component from what I can tell is to provide "art direction" and visual styling to an image (border, radius, block, shadow, etc...). We should keep this as low cost as possible IMO and mapping towards something like text would give us more benefit. 
Consider this approach:
```html
<fluent-image border shape="circle"><img src="/foo" alt="foo image" /></fluent-image>
```
This allows us to do *both* - we can provide the intent of the component and provide encapsulation for the visual styles for the image, while leveraging _more_ of the platform and not having to enumerate upwards of twenty or so attributes just to render an image.

chrisdholt commented on packages/web-components/src/image/image.template.ts:
```diff
@@ -0,0 +1,23 @@
+export const template: ElementViewTemplate<Image> = html<Image>`
+  <template>
+    <img
+      alt="${x => x.alt}"
+      block="${x => x.block}"
+      border="${x => x.border}"
+      borderRadius="${x => x.borderRadius}"
+      fit="${x => x.fit}"
+      margin="${x => x.margin}"
+      role="${x => (x.presentation ? 'presentation' : null)}"
+      shadow="${x => x.shadow}"
+      shape="${x => x.shape}"
+      src="${x => x.src}"
```
Were we to take an approach like this in the future, I think you want to bind any of the attributes which aren't native to the img element on the host.

chrisdholt commented on packages/web-components/src/image/image.ts:
```diff
@@ -0,0 +1,101 @@
+   */
+  @attr({ mode: 'boolean' })
+  public border?: boolean;
+  /**
+   * Image margin
+   *
+   * @public
+   * @remarks
+   * HTML attribute: margin.
+   */
+  @attr({ mode: 'boolean' })
+  public margin?: boolean;
```
Is this part of our public API or is it additive? Not seeing it on the Fluent v9 storybook

halerankin replied:
Nope, this is one prop that's not present in the Fluent v9 react component. I will add a note to the docs file stating where this component differs from the Fluent v9. That and border-radius look like the only two extras.

halerankin replied:
Do you want me to make adjustments here?

eljefe223 replied:
Do we think `margin` is the best name for what this is doing? Sure it adds a 16px margin around the host but it seems like an odd name as its a boolean.

chrisdholt replied:
@halerankin I don't see this on the v9 image component - am I missing it?

halerankin replied:
Here's the Figma spec I was given prior to building it out: https://www.figma.com/file/05wt6TAsEmgsCVZfPrpcWx/Image?node-id=0%3A1&t=KZwLafEjmVs2M7AG-0

chrisdholt replied:
Sure - I think we need clarity here...though I'd argue if the intent is to add margin to the component, this is 100% the wrong way given there is additional cost of creating an attribute, adding bindings and lifecycle, etc. We should just use CSS in this case.

chrisdholt replied:
Here's the latest v9 - just not seeing margin as enumerated as a prop, so I don't think we should move forward with it here (both for that reason and the one above)

halerankin replied:
So, do we just leave it up to the dev implementing this component to add the 16px margin when needed?

chrisdholt replied:
Absolutely - I think my question would be, "Why is margin special as far as CSS goes where it gets an attribute and others do not?"

Similarly, I don't see where we're actually doing anything *with that*.

halerankin replied:
Looking at the spec, I see a prescribed layout: optional margin of 16px. We have no way to strictly require this 16px value. We *could* by adding a CSS class in our styles that provides that 16px but outside that we have zero control over margin value.

chrisdholt replied:
I just looked at the spec and I see where the confusion is - I'm going to maintain that I don't think that is intended as part of the component API as far as attributes or properties are concerned. In the case that it's added to the react API, I'd probably hold on my stance given that margin should not be treated as a CSS property in a JS API given that we can easily apply it via CSS with zero cost. Were we to set this in the class as an attribute, we would need to then map it to css _from_ the JS class which would either be a unique token (odd, expensive) or an inline style (odd, unintuitive). 

To my point above, I think we need to understand why margin deserves special treatment over something like "padding". Further, what if I need margin-inline or just margin-block? Is the only method I have to set this setting a singular value for top/right/left/bottom? If so, we're likely creating confusion because...are those not supported values? If they are supported, I likely set them via CSS, so we return again to why margin is specially set for this scenario.

halerankin replied:
In the spirit of keeping things simple, I think omitting margin as a component class property is best. Leaving the margin to the web component consumer provides them with more flexibility.

chrisdholt commented on packages/web-components/src/image/image.stories.ts:
```diff
@@ -0,0 +1,245 @@
+    <fluent-image
+      ?block=${x => x.block}
+      ?border=${x => x.border}
+      borderRadius=${x => x.borderRadius}
```
I'd update this to match the attribute (instead of the property) as it's an attribute on the element.
```suggestion
      border-radius=${x => x.borderRadius}
```

chrisdholt commented on packages/web-components/src/image/image.options.ts:
```diff
@@ -0,0 +1,46 @@
+ * @public
+ */
+export const ImageFit = {
+  none: 'none',
+  center: 'center',
+  contain: 'contain',
+  cover: 'cover',
+  default: 'default',
```
@miroslavstastny just to call out another oddity IMO - having "none" here seems odd if this is going to be nullable (meaning we set defaults in CSS). Is there someone we can sync with to understand the intention of having both "none" and "default" as values in the types here? I see that "none" still applies object fit properties, so that makes sense. But for scenarios where "none" is undefined, `default` seems odd to enumerate given it won't do anything.

No action here, just stating that it's confusing from an API standpoint especially if we don't set it.

halerankin replied:
I should have called this out when I first wrote it in response to the Figma spec. It's does seem silly to have both `none` and `default`. The developer should get the default and opt in for `center`, `contain` and `cover`. Same for image fit: square is default, so only `circular` and `rounded` should be available.

halerankin replied:
Or do we have to be so explicit?

halerankin replied:
Any actionable item here?

chrisdholt replied:
I don't think so but it's odd and want to raise awareness in these cases.

chrisdholt commented on packages/web-components/src/image/image.styles.ts:
```diff
@@ -0,0 +1,78 @@
+ * @public
+ */
+export const styles = css`
+  :host ::slotted(*) {
```
general question here as I re-review - do we want to support * (all) slotted selectors here or just img/svg, or does this also actually work with `<picture></picture>`? Perhaps we should start with just `img`? Curious of others thoughts here. `*` selector is costly but if we're resolving any element it makes sense. In the case of `img`, do we just want `img` given that's what is returned in v9?

halerankin replied:
I'm good with img.

halerankin replied:
Do I write `::slotted(img)` ?

chrisdholt commented on packages/web-components/src/image/image.styles.ts:
```diff
@@ -0,0 +1,78 @@
+  :host([shape='circular']) ::slotted(*) {
+    border-radius: 100%;
+  }
+  :host([shape='rounded'][borderRadius='small']) ::slotted(*) {
```
@halerankin for `borderRadius` we should use attribute syntax `border-radius`

halerankin replied:
Done.

halerankin commented:
@chrisdholt Updated code. I removed `alt`, `role` and `src` from the component as they are already native attributes on the img element. I removed the block attribute because the same results can be achieved by setting `fit` to `contain`. I removed `margin` attribute based on our conversation. I updated the args table to show that these can all be supplied by the component consumer.

fabricteam commented:
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/0b8e596e512b17f42d382be384d9c0df9c43cc0b'>0b8e596e512b17f42d382be384d9c0df9c43cc0b</a></sub>
<div id="bundleSizeComment9423"/>

chrisdholt commented on packages/web-components/src/image/image.options.ts:
```diff
@@ -0,0 +1,46 @@
+ * Border radius
+ * @public
+ */
+export const BorderRadius = {
```
@halerankin should this be namespaced for Image since it is specific to image? Sorry for missing this!

chrisdholt commented on packages/web-components/src/image/image.options.ts:
```diff
@@ -0,0 +1,46 @@
+ * Image fit
+ * @public
+ */
+export const ImageFit = {
```
Is it possible that other components have `ImageFit` (@miroslavstastny do you think this needs namespacing or are we safe here?) It is the fit of the image....

chrisdholt replied:
to be clear here, I think this is find, but I want to make sure we don't need to be doing `ImageImageFit` :/

chrisdholt replied:
Closing this as I've answered my own - attribute is `fit`, so this is already namespaced. Move along.

halerankin replied:
These are scoped to this component. How will these named attributes/types impact other components whose attributes are named the same?

chrisdholt replied:
No change to this one, only to `BorderRadius` above

chrisdholt commented on packages/web-components/src/index.ts:
```diff
@@ -3,5 +3,5 @@ export * from './counter-badge/index.js';
 export * from './text/index.js';
-
 export * from './theme/index.js';
+export * from './image/index.js';
```
@halerankin lets be sure we alphabetize this - the pain will primarily be up front. This won't be as difficult when we get most components in :)

chrisdholt reviewed (approved):
Approving but I'd like to make sure that @miroslavstastny has a chance to confirm he's good with this, and a couple of things to clean up.

miroslavstastny reviewed (changes requested):
The PR currently contains invalid changes, perhaps caused by merge/rebase.

halerankin commented:
@miroslavstastny I will open a new PR with only my changes.
